pre_commit/languages/golang.py

```diff
@@ -20,82 +24,139 @@
 health_check = helpers.basic_health_check
 
 
-def get_env_patch(venv: str) -> PatchesT:
+@functools.lru_cache(maxsize=1)
+def get_default_version() -> str:
+    if helpers.exe_exists('go'):
+        return 'system'
+    else:
+        return '1.19.4'  # C.DEFAULT
+
+
+def get_env_patch(
+        venv: str,
+        language_version: str,
+) -> PatchesT:
+    if language_version == 'system':
+        return (
+            ('PATH', (os.path.join(venv, 'bin'), os.pathsep, Var('PATH'))),
+        )
+
     return (
-        ('PATH', (os.path.join(venv, 'bin'), os.pathsep, Var('PATH'))),
+        ('GOROOT', os.path.join(venv, '.go')),
+        ('GOPATH', venv),
+        (
+            'PATH', (
+                os.path.join(venv, 'bin'), os.pathsep,
+                os.path.join(venv, '.go', 'bin'), os.pathsep, Var('PATH'),
+            ),
+        ),
     )
 
 
-@contextlib.contextmanager
-def in_env(prefix: Prefix) -> Generator[None, None, None]:
-    envdir = prefix.path(
-        helpers.environment_dir(ENVIRONMENT_DIR, C.DEFAULT),
-    )
-    with envcontext(get_env_patch(envdir)):
-        yield
+def _get_arch() -> str:  # pragma: no cover
+    arch = platform.machine().lower()
+    if arch in ('x86_64',):
+        return 'amd64'
+    if arch in ('i386',):
+        return '386'
+    if arch in ('armv6l', 'armv7l'):
+        return 'armv6l'
+    if arch in ('arm64', 'aarch64', 'armv8'):
+        return 'arm64'
+    return arch
 
 
-def guess_go_dir(remote_url: str) -> str:
-    if remote_url.endswith('.git'):
-        remote_url = remote_url[:-1 * len('.git')]
-    looks_like_url = (
-        not remote_url.startswith('file://') and
-        ('//' in remote_url or '@' in remote_url)
+def _get_url(version: str) -> str:
+    os_name = platform.system().lower()
+    ext = 'zip' if os_name == 'windows' else 'tar.gz'
+    return (
+        f'https://dl.google.com/go/go{version}.{os_name}-{_get_arch()}.{ext}'
     )
-    remote_url = remote_url.replace(':', '/')
-    if looks_like_url:
-        _, _, remote_url = remote_url.rpartition('//')
-        _, _, remote_url = remote_url.rpartition('@')
-        return remote_url
+
+
+def _install_go(prefix: Prefix, version: str, dest: str) -> None:
+    try:
+        resp = urllib.request.urlopen(_get_url(version))
+    except urllib.error.HTTPError as exec:  # pragma: no cover
+        if exec.code == 404:
+            raise ValueError(
+                'Could not find a version matching your system requirements '
+                f'(os={platform.system().lower()}; arch={_get_arch()})',
+            ) from exec
+        else:
+            raise
     else:
-        return 'unknown_src_dir'
+        with tempfile.NamedTemporaryFile(delete=False) as tmp_file:
+            shutil.copyfileobj(resp, tmp_file)
+            shutil.unpack_archive(
+                tmp_file.name,
+                dest,
+                format='zip' if sys.platform == 'win32' else 'gztar',
+            )
```

Comment:
hmm I thought the whole point of unpack_archive was that it's automatic?

Reply:
That's what I thought at first, but when not provided, the format is only inferred using the filename extension.

Reply:
should be easy to preserve the original extension right?

Reply:
The only way I can think of is to add it as a suffix to the named temp file (basically move the whole ternary there)

Reply:
the url has the extension, no?

Reply:
I've suffixed the whole name, might be useful for someone someday inspecting the temp dir :D